### PR TITLE
Adding toolbar to the Image Gallery component

### LINKED ISSUE #34437: [Components] Image gallery toolbar
Once https://github.com/woocommerce/woocommerce/pull/34435 is merged, we should add a toolbar to adjust image gallery items.  This toolbar should allow:

* Dragging of the item via a handle
* Updating the order of an item up/down
* Setting the cover for the gallery
* Replacing an image (should open the media modal)
* Removing an item

Some additional props for the `ImageGallery` will be useful:

* `onRemove`
* `onReplace`
* `onOrderChange`

Image

### PR DESCRIPTION
**Pull request by joelclimbsthings**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adding the Image Gallery toolbar.

Closes #34437 .

### How to test the changes in this Pull Request:

1. Run storybook and browse to Image Gallery -> Basic story
2. Enable console and view callback logs while testing basic functions:
     - Drag and drop item
     - Click item to reveal toolbar
     - Toolbar -> move left
     - Toolbar -> move right
     - Toolbar -> drag item by handle
     - Toolbar -> set item as cover image
     - Toolbar -> replace item
     - Toolbar -> remove item
4. Ensure that you cannot drag/drop cover item or move with toolbar
5. Ensure that you cannot move final item forward

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: packages/js/components/src/sortable/sortable-item.tsx
Comment (joelclimbsthings) [inline packages/js/components/src/sortable/sortable-item.tsx:74]: Necessary to prevent Firefox from redirecting to the image itself.

File: packages/js/components/src/image-gallery/icons/cover-image.js
Comment (joshuatf) [inline packages/js/components/src/image-gallery/icons/cover-image.js]: Nitpicky, but `CoverImageIcon` maybe to reduce ambiguity?

File: packages/js/components/src/sortable/sortable.tsx
Comment (joshuatf) [inline packages/js/components/src/sortable/sortable.tsx]: Any way we could avoid using a class name?  If not, maybe we could create an HOC that passes the classname to the child?

```
<NonSortableItem>
  // Item that can't be sorted
</NonSortableItem>
```

This prevents the consumer from knowing too much about specific classes and might make things a little less brittle.

File: packages/js/components/src/sortable/sortable-item.tsx
Comment (joshuatf) [reply packages/js/components/src/sortable/sortable-item.tsx:74]: Nice!

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joshuatf) [inline packages/js/components/src/image-gallery/image-gallery.tsx]: I think `moveIndex` will need to be imported from `../sortable/utils`.

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joshuatf) [inline packages/js/components/src/image-gallery/image-gallery.tsx]: I know you're not finished with this yet, but it would be great to extract the toolbar logic to a separate component.

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joshuatf) [inline packages/js/components/src/image-gallery/image-gallery.tsx]: I think what's happening here is that when you reorder, it's passing back the children exactly as they are.  In your case, it looks like the children might be:

```
<ImageGalleryItem />
<SortableHandle><ImageGalleryItem /></SortableHandle>
<SortableHandle><ImageGalleryItem /></SortableHandle>
<SortableHandle><ImageGalleryItem /></SortableHandle>
...
```

On the next render after you've updated state with the reordered items, when you check the item, even though `isCoverImage` is `true`, the first item would look like this:

```
<SortableHandle isCover><ImageGalleryItem /></SortableHandle>
```

I have a few recommendations in order of what I would consider highest priority:

1. Is it possible just to make the first item `isCover`?
3. Instead of tracking the whole child as the `coverImage`, I think we should just track a `key` or ID of some sort
4. `SortableHandle` is probably best moved inside the `ImageGalleryItem` to make it a little easier for you to pass props around
5. Optionally, we could pass props from `SortableHandle` down to any children it has.

File: packages/js/components/src/image-gallery/icons/cover-image.js
Comment (joelclimbsthings) [reply packages/js/components/src/image-gallery/icons/cover-image.js]: Done!

File: packages/js/components/src/sortable/sortable.tsx
Comment (joelclimbsthings) [reply packages/js/components/src/sortable/sortable.tsx]: I'm not sure an HoC approach would be the best here given the issues with wrapping `children` with any components; it could encourage behavior that will cause issues (as I did earlier). Also the individual child that this will apply to will be rotating and changing due to interactions.

For now I switched it to the `notSortableIndexes` prop within the `<Sortable />` component, although this is a little rigid as it only accepts indexes. We could make it more flexible by allowing it to accept child `keys` as well, although this seems like a awkward/rare use case in general.

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joelclimbsthings) [reply packages/js/components/src/image-gallery/image-gallery.tsx]: I had added it to the exports for the base package, but it's easy enough to change this back if that's not preferred. Let me know.

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joelclimbsthings) [reply packages/js/components/src/image-gallery/image-gallery.tsx]: This has all been refactored pretty thoroughly. I've still left some of the basic functions within this component (`moveItem`/etc), as I've considered it the layer in which we're managing the image gallery `children` in general, move more niche behavior (ie `setAsCover`) to the toolbar component.

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joelclimbsthings) [reply packages/js/components/src/image-gallery/image-gallery.tsx]: Thanks for this @joshuatf , the issue with the nested children was my primary issue since I wasn't looking in the right places. Appreciate it!

> Is it possible just to make the first item isCover?

This has been my approach, so I'm unsure if you're talking about something different. The item under the first index (`0`) is always set as the cover. Let me know if I'm misunderstanding something.

> Instead of tracking the whole child as the coverImage, I think we should just track a key or ID of some sort \

This had been my intention but I had mystery keys appearing in unpredictable ways, and only too late for me to utilize them. I now believe this was stemming from the `Children.map()` function(s) which automaticly add them, and I've migrated this to set my own keys on the base children as they're initialized and tracking them with those. Seems to work! 👍🏻 

> SortableHandle is probably best moved inside the ImageGalleryItem to make it a little easier for you to pass props around

Yup, good call and I've moved all wrapping functions within `ImageGalleryItem`. 

> Optionally, we could pass props from SortableHandle down to any children it has.

Doesn't seem necessary after the refactors, but glad to throw this in if desired.

**Comment by joelclimbsthings** [comment]
Thanks for the preliminary review  @joshuatf ! I've been able to resolve the remaining issues and I've responded to your comments above. 

One thing of note, I was finding it pretty tedious and bug-prone to pass the `onDragStart` and `onDragEnd` around multiple components to get them to the `<SortableHandle />`'s. I've experimentally [replaced those with context](https://github.com/woocommerce/woocommerce/pull/34871/commits/09085289c8cd460bf316de7f499da037a0a3b8ca), which operates behind the scenes within `<Sortable />` and makes it work without extra props as long as it's a descendant of the `<SortableItem />`; I believe this should always be the case, but let me know if this could miss something. Feels much cleaner to me so far.

It's somewhat a breaking change just in that it could result in TS errors, but I believe I've addressed those.

I still need to address accessibility concerns, but may advocate getting that done in a separate PR (glad to work on this next).

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joshuatf) [reply packages/js/components/src/image-gallery/image-gallery.tsx]: >This has been my approach, so I'm unsure if you're talking about something different. The item under the first index (0) is always set as the cover. Let me know if I'm misunderstanding something.

I don't think you'll need to track the `coverImage` in state in this case.  You can just grab the first item from the list and pass the `isCover` prop to it, right?

Comment (joshuatf) [reply packages/js/components/src/image-gallery/image-gallery.tsx]: Scratch that.  Just looked at the latest iteration and it does not track cover image in state.

File: packages/js/components/src/sortable/types.ts
Comment (joshuatf) [inline packages/js/components/src/sortable/types.ts:8]: ❤️  This is awesome

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joshuatf) [inline packages/js/components/src/image-gallery/image-gallery.tsx]: I think a compositional API might be slightly advantageous over props here.  Could we make a component `NonSortableItem` or something to that effect that we can wrap items with?

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joshuatf) [inline packages/js/components/src/image-gallery/image-gallery.tsx]: It would be really great to have `moveItem`, `removeItem`, and `replaceItem` extracted to utils with some unit tests.

The state methods would need to be replaced with callbacks in order for this to work.

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joshuatf) [inline packages/js/components/src/image-gallery/image-gallery.tsx]: Maybe `isToolbarVisible`?

File: packages/js/components/src/image-gallery/image-gallery-toolbar.tsx
Comment (joshuatf) [inline packages/js/components/src/image-gallery/image-gallery-toolbar.tsx]: We'll need to wrap this and the below labels in i18n calls.

```suggestion
				label={ __( 'Options', 'woocommerce' ) }"
```

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joshuatf) [inline packages/js/components/src/image-gallery/image-gallery.tsx]: `activeToolbarIndex`?

**Review by joshuatf — COMMENTED** [review]
Nice work, Joel!  Looking good and love the context additions for the sortable handle.

Added a few clean-up comments.  I think we should probably also hide the toolbar on blur, right?  It seems to stay active even when clicking away:

Screen Shot 2022-10-03 at 10 53 59 AM

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joelclimbsthings) [reply packages/js/components/src/image-gallery/image-gallery.tsx]: Sounds good @joshuatf , this was possible by utilizing classes again (so we don't mandate a particular prop in child components) and moving it into the `ImageGalleryItem` component. 

Done in https://github.com/woocommerce/woocommerce/pull/34871/commits/46e5c7b1e1ce90586e62ab75fa6392a97c854a28.

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joelclimbsthings) [reply packages/js/components/src/image-gallery/image-gallery.tsx]: `moveItem` doesn't add any logic to `moveIndex`, so I've left that one out. I've moved the others into the `utils.ts` file and I'm adding tests now.

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joelclimbsthings) [reply packages/js/components/src/image-gallery/image-gallery.tsx]: 👍🏻

File: packages/js/components/src/image-gallery/image-gallery-toolbar.tsx
Comment (joelclimbsthings) [reply packages/js/components/src/image-gallery/image-gallery-toolbar.tsx]: Whoops, forgot that again. Done.

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joelclimbsthings) [reply packages/js/components/src/image-gallery/image-gallery.tsx]: 👍🏻

**Comment by joelclimbsthings** [comment]
Thanks for the thorough review @joshuatf ! I've implemented your suggestions. I'm currently still working on adding tests, but running into issues with my local in actually running them. I'd like to get the other changes out to you for review in the mean time.

**Comment by AnnaMag** [comment]
@joelclimbsthings is it an option to add tests in a follow-up and merge these changes to unblock the work for this section?

**Comment by joelclimbsthings** [comment]
@AnnaMag That's fine by me, as I'm planning a follow-up to address more accessibility concerns as well. I'll let @joshuatf chime in on his take

Edit: I managed to get it working and add a few tests this morning 👍🏻

File: packages/js/components/src/image-gallery/image-gallery.tsx
Comment (joshuatf) [inline packages/js/components/src/image-gallery/image-gallery.tsx]: Nitpicky, but would be nice if these names matched:
```suggestion
								'is-toolbar-visible': isToolbarVisible,
```

File: packages/js/components/src/image-gallery/image-gallery-item.tsx
Comment (joshuatf) [inline packages/js/components/src/image-gallery/image-gallery-item.tsx:33]: I'm wondering if this might be overly opinionated for this component and something that we should let the consumer decide (whether cover items should or should not be sortable).

Just thinking out loud; we can save this for a later date.

File: packages/js/components/src/image-gallery/test/utils.tsx
Comment (joshuatf) [inline packages/js/components/src/image-gallery/test/utils.tsx:31]: Thanks for these tests! 🎉

**Review by joshuatf — DISMISSED** [review]
Changes look great!  Thanks for the tests as well 💯 

Left one minor naming comment, but otherwise I think if we have further tweaks we can address in a follow-up.

File: packages/js/components/src/image-gallery/image-gallery-item.tsx
Comment (joelclimbsthings) [reply packages/js/components/src/image-gallery/image-gallery-item.tsx:33]: Valid thought, and sounds good on my end 👍🏻

**Review by joshuatf — APPROVED** [review]
Thanks for the class name change!  LGTM 🚢

**Pull request merged**
